Profile screen

DarkRavenn opened the pull request:
### 1/3 часть
Архитектура: **MVP** 
Верстка: **Кодом**
[Таск трекер](https://github.com/users/DarkRavenn/projects/3/views/1)
[Карточка по этой части](https://github.com/DarkRavenn/FakeNFT/issues/2)

DarkRavenn committed:
Добавлен файл с описанием декомпозиции модуля профиль

DarkRavenn committed:
Добавлен таб профиль

DarkRavenn committed:
Добавлена модель UserProfile

DarkRavenn committed:
Добавлены мок данные в mockUserProfile

DarkRavenn committed:
Сверстан и заполнен моковыми данными аватар, имя, описание, сайт

DarkRavenn committed:
Реализована UITableView с ячейками "Мои NFT", "Избранные NFT", "О разработчике"

DarkRavenn committed:
Добавлена обработка нажатий на ячейки таблицы (переход на соответствующие экраны) и Кнопки в нав баре

DarkRavenn committed:
Добавлен экран и функционал редактирования профиля

DarkRavenn committed:
Переписан экран профиль на получение и обработку данных с сервера

DarkRavenn committed:
Добавлена загрузка изображения с помощью KIngFisher

DarkRavenn committed:
Рефакторинг: Организовал код с MARK для улучшения читаемости

- Переместил свойства, инициализаторы и методы в соответствующие секции MARK
- Разделил @objc private методы и другие приватные методы

DarkRavenn committed:
Рефакторинг кода с учетом архитектуры MVP

Helvetii951 commented on FakeNFT/Scenes /Profile/EditProfilePresenter.swift:
```diff
@@ -0,0 +1,39 @@
+    func updateProfile(profile: UserProfile)
+}
+
+class EditProfilePresenter: EditProfilePresenterProtocol {
```
Рекомендую не забывать проставлять `final`.

Использование final может помочь компилятору Swift оптимизировать код, так как он знает, что метод или класс не будет переопределен или унаследован. Это может привести к более быстрому выполнению кода, так как компилятор может использовать прямые вызовы методов вместо динамической диспетчеризации.

Про диспетчеризацию, если интересно, можно почитать здесь:
https://habr.com/ru/post/673400/
https://habr.com/ru/company/simbirsoft/blog/673636/

Helvetii951 commented on FakeNFT/Scenes /Profile/EditProfilePresenter.swift:
```diff
@@ -0,0 +1,39 @@
+class EditProfilePresenter: EditProfilePresenterProtocol {
+    // MARK: - Private Properties
+    private weak var view: EditProfileControllerProtocol?
+    private let servicesAssembly: ServicesAssembly
+    
+    // MARK: - Initializers
+    init(view: EditProfileControllerProtocol, servicesAssembly: ServicesAssembly) {
+        self.view = view
+        self.servicesAssembly = servicesAssembly
+    }
```
Презентеру очень желательно не знать о `ServicesAssembly`. Он должен получить уже готовую зависимость.
```swift
init(view: EditProfileControllerProtocol, profileService: ProfileService) { ... }

....

presenter = EditProfilePresenter(view: self, profileService: servicesAssembly.profileService)
```

Helvetii951 commented on FakeNFT/Scenes /Profile/FavoritesNFT.swift:
```diff
@@ -0,0 +1,35 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        view.backgroundColor = UIColor.white
+        navigationItem.title = NSLocalizedString("Profile.FavoritesNFT.title", comment: "")
```
Рекомендую всегда генерировать строки. Есть два подхода:

1) Использовать библиотеки [R.Swift ](https://github.com/mac-cain13/R.swift)или [SwiftGen](https://github.com/SwiftGen/SwiftGen).
2) Генерить [с помощью xCode](https://developer.apple.com/documentation/xcode/localizing-and-varying-text-with-a-string-catalog).

Helvetii951 commented on FakeNFT/Scenes /Profile/MyNFT.swift:
```diff
@@ -0,0 +1,35 @@
+
+import UIKit
+
+final class MyNFT: UIViewController {
```
Нэйминг `MyNFT` не очень хорошо отражает назначение сущности. `MyNFTController` подошло бы гораздо лучше :)

Helvetii951 commented on FakeNFT/Scenes /Profile/ProfileController.swift:
```diff
@@ -0,0 +1,269 @@
+        view.backgroundColor = .systemBackground
+        
+        view.addSubview(contentView)
+        
+        presenter = ProfilePresenter(view: self, servicesAssembly: servicesAssembly)
+        
```
Очень хорошая идея объявить презентер протоколом, не очень хорошая - создавать в самом контроллере конкретную реализацию.

Для гибкости и тестируемости кода желательно `presenter ` как-нибудь инжектить.

[Неплохая статья по внедрению зависимостей](https://habr.com/ru/companies/tinkoff/articles/546360/).

Helvetii951 commented on FakeNFT/Services/ProfileService.swift:
```diff
@@ -0,0 +1,89 @@
+    }
+    
+    func loadProfile(completion: @escaping ProfileCompletion) {
+        if let profile = storage.getProfile() {
+            completion(.success(profile))
+            print("Загружено: UserDefaults")
+            return
+        }
```
Допустим, у нас есть приложение и веб-версия. Если поменяем профиль в веб-версии, приложение об этом никак не узнает, пользователь будет видеть старую версию данных. Есть два способа решить эту проблему:

1. Установить время жизни сохраненных данных. Т.е., если после сохранения прошло, например 10 минут, очищаем хранилище и качаем профиль заново.
2. Никогда не хранить скаченный профиль.

Helvetii951 commented on FakeNFT/Services/ProfileService.swift:
```diff
@@ -0,0 +1,89 @@
+            return
+        }
+        
+        UIBlockingProgressHUD.show()
+        let request = ProfileRequest()
+        print("""
+              method: \(request.httpMethod)
+              endpoint: \(request.endpoint!)
+              \n
+              """)
+        networkClient.send(request: request, type: UserProfile.self) { result in
+            UIBlockingProgressHUD.dismiss()
```
Сервис не должен заниматься работой с UI. У него другая обязанность. HUD необходимо перенести в UI слой.

Helvetii951 commented on FakeNFT/Scenes /Profile/ProfileController.swift:
```diff
@@ -0,0 +1,269 @@
+    
+    // MARK: - View Life Cycles
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        view.backgroundColor = .systemBackground
+        
+        view.addSubview(contentView)
+        
+        presenter = ProfilePresenter(view: self, servicesAssembly: servicesAssembly)
```
Лучше не делать лишние переносы строк, они могут затруднять чтение кода.

Дополнительный материал про кодстайл:  

- [Подробное описание codestyle](https://google.github.io/swift/) от Google
- [От Ray Wanderlich](https://github.com/raywenderlich/swift-style-guide)
- [На русском от команды tutu](https://github.com/tutu-ru/swift-style-guide)

DarkRavenn committed:
Правки по ревью:

- Проставлен пропущенный final у классов
- Добавлено получение готовой зависимости презентерам
- Исправлено название ex классов FavoritesNFT и MyNFT
- Добавлен DI для EditProfilePresenter и EditProfilePresenter
- Добавлен таймер кеша профиля
- HUD перенесен из presenter в controller
- Убран лишний пробел в ProfileController

DarkRavenn committed:
Правки по ревью:

-Добавил генерацию строк через Xcode

DarkRavenn committed:
Правки по ревью:

- Удалены мок фото
- Изменена логика замены ссылки при нажатии на фото

DarkRavenn commented:
Ревью: 
не смог воспроизвести: "2. После обновления профиля ссылка становится серой. [Скрин](https://postimg.cc/qhYdrDsB)." 

если еще воспроизводится дай пожалуйста больше деталей по воспроизведению в идеале с видео.

Helvetii951 commented:
> Ревью: не смог воспроизвести: "2. После обновления профиля ссылка становится серой. [Скрин](https://postimg.cc/qhYdrDsB)."
> 
> если еще воспроизводится дай пожалуйста больше деталей по воспроизведению в идеале с видео.

Привет! Вот [видео](https://vkvideo.ru/video8050623_456239211?list=ln-wowzq4na4cnRBCHzZX).

DarkRavenn committed:
Исправлена серая ссылка после обновления профиля
